### ISSUE TITLE
Questions on using accessors for restricted accesses

### ISSUE THREAD

**Issue by roastduck** [kokkos/mdspan#169]
Hi, in Section 2.7 of the proposal, I see one of the design goal of custom accessors is to apply `restrict` (non-aliasing hint) to memory accesses (if a C++ implementation supports it). But I have not figured out how to implement such an accessor. Can someone provide an example?

The problem is that I cannot simply return a `restrict`ed pointer *for each item* in a `mdspan`, because the semantics I want is a `mdspan` `a` is never an alias of another `mdspan` or ordinary pointer `b`, but items inside `a` can be aliases, for example `a[3]` and `a[i]` can be aliases.

**Comment by mhoemmen** [comment]
Greetings and thanks for your interest!  Your use case is different than what we envisioned for a "no-aliasing" accessor.  May we ask what data types you have in mind for the elements of the array `a`?

I'm not a C expert, but I'm not 100% convinced that your use case is not undefined behavior.  I think it might be legal use of C `restrict`, because `a+3` and `a+i` are both "based on" `a` (in the language of the C Standard), but it's definitely not what we had in mind.

There are only two places one could apply `restrict` (or `__restrict__`, `__restrict`, etc.) to a custom accessor:

1. to the accessor's `data_handle_type`, or
2. to the accessor's `reference`.

Regarding (1), the resulting custom accessor would look almost like `default_accessor`, except that `data_handle_type` would be a `restrict` pointer.  The offset policy would actually be `default_accessor` (so that offset access is not marked `restrict`, since it's aliasing access).  Assuming that what you want to accomplish is compatible with `restrict`, this is probably what you would want to do.

Regarding (2), some compilers, [such as GCC](https://gcc.gnu.org/onlinedocs/gcc/Restricted-Pointers.html), permit applying this keyword to references.  However, I don't think this would accomplish your goal.  If the accessor's `reference` type is marked `restrict`, then the reference can't alias another reference in the local context.  The accessor has no way of knowing whether it will be used in code that refers to both `a[3]` and `a[i]`.  If _you_ know that you won't be doing that, then you can always apply `restrict` to the reference by hand in the code.

In summary: write a custom accessor that looks almost like `default_accessor`, except that `data_handle_type` is a `restrict` pointer.  Make sure that the offset policy does _not_ use `restrict`.

**Comment by roastduck** [comment]
Thanks! It solved my problem. Actually I have to separate `data_handle_type` form `default_accessor::offset`'s parameter and return value. The parameter should be `restrict` while the return value should not.

By the way, I guess my use case is quite common: people often distinguish different arrays, but not elements in one array. For example, in the following code

```
a[i] += b[i]
a[i] += b[i]
```

We can guarantee `a` and `b` are different from time to time, but the two `a[i]`s are always the same.

**Comment by mhoemmen** [comment]
@roastduck Thanks for explaining your use case, and for letting us know that the suggestion worked!  It would be excellent if you could share any performance improvements you see from using this accessor -- please don't feel obligated, but we're just curious : - ) .

**Comment by roastduck** [comment]
Sorry but I am still confused after digging into the details. The effect of a `restrict` hint is that it guarantees a pointer has no alias *during its lifetime* (https://en.wikipedia.org/wiki/Restrict). But if I mark `data_handle_type` of `default_accessor::offset`'s and `default_accessor::access`'s parameter as above, it's lifetime ends when the functions return. I'm afraid it will have no effect if I use `restrict` in this way.

Suppose a `mdspan` can read the type from `accessor::data_handle_type` and use it as the pointer type inside the `mdspan`, it may solve the problem. In other words, maybe we can make `accessor::data_handle_type` an public interface, or requirement, of an `accessor`.

**Comment by roastduck** [comment]
Another problem is: when directly accessing a `restrict` pointer `a` with an offset, like `a[i]`, there is no other pointer involved, and we are accessing *based on* a. However, in the `mdspan` case, we need to first convert `a` to a non-`restrict` pointer or reference in `accessor::offset` or `accessor::access`. I'm not sure whether this will stop compilers from treating the resulting pointer or reference as non-aliasing.

**Comment by mhoemmen** [comment]
@roastduck wrote:
> Suppose a mdspan can read the type from `accessor::data_handle_type` and use it as the pointer type inside the mdspan, it may solve the problem.

`mdspan::data_handle_type` is an alias to its accessor type's `data_handle_type`.  They are the same type.  There is no `pointer` type any more, btw (it's called `data_handle_type` now).

I haven't tested the following, but here's the accessor type that I had in mind.
```c++
template<class ElementType>
  struct restrict_accessor {
    using offset_policy = std::default_accessor<ElementType>;
    using element_type = ElementType;
    using reference = ElementType&;
    using data_handle_type = ElementType* restrict;

    constexpr restrict_accessor() noexcept = default;

    template<class OtherElementType>
    requires(std::is_convertible_v<OtherElementType(*)[], element_type(*)[]>)
    constexpr restrict_accessor(restrict_accessor<OtherElementType>) noexcept {}

    constexpr reference access(pointer p, size_t i) const noexcept {
        return p[i];
    }
    constexpr typename offset_policy::data_handle_type offset(data_handle_type p, size_t i) const noexcept {
        return p + i;
    }
  };
}
```

> But if I mark data_handle_type of `default_accessor::offset`'s and `default_accessor::access`'s parameter as above, it's lifetime ends when the functions return.

https://en.cppreference.com/w/c/language/restrict gives examples of the kinds of optimizations that may apply to `restrict`.  For example, "[r]estricted pointers can be assigned to unrestricted pointers freely, the optimization opportunities remain in place as long as the compiler is able to analyze the code:"
```c++
void f(int n, float * restrict r, float * restrict s) {
   float * p = r, * q = s; // OK
   while(n-- > 0) *p++ = *q++; // almost certainly optimized just like *r++ = *s++
}
```
This is relevant because `restrict_accessor::reference` is _not_ `restrict`.  (It's not formally correct to apply C `restrict` wording to C++ references, but I'll assume in what follows that this has a reasonable definition.)  The idea is that even though `p[i]` has type `ElementType&` and not `ElementType& restrict`, the compiler can figure out that the reference comes from a pointer based on `p`, which is marked `restrict`.

> I'm not sure whether this will stop compilers from treating the resulting pointer or reference as non-aliasing.

This can only be determined by experiment.  Compilers are not required to do anything with `restrict`.  Any use of this keyword is not Standard C++, so you'll have to refer to the compiler's documentation, look at the assembler output, and do performance experiments.

**Comment by roastduck** [comment]
Thank you again for your answers. It seems to work in principle, so I did some experiments with g++ 11.2.0 in my local machine supporting AVX2 vector instructions to see if the `restrict` hint affecting vectorizing.

I wrote 4 versions of `vadd` in the following code:

```
#include "mdspan.hpp"

#include <cassert>
#include <iostream>

namespace stdex = std::experimental;

template<class ElementType> struct restrict_accessor {
    using offset_policy = stdex::default_accessor<ElementType>;
    using element_type = ElementType;
    using reference = ElementType&;
    using data_handle_type = ElementType* __restrict__;

    constexpr restrict_accessor() noexcept = default;

    template<class OtherElementType>
    requires(std::is_convertible_v<OtherElementType(*)[], element_type(*)[]>)
    constexpr restrict_accessor(restrict_accessor<OtherElementType>) noexcept {}

    constexpr reference access(data_handle_type p, size_t i) const noexcept {
        return p[i];
    }
    constexpr typename offset_policy::data_handle_type offset(data_handle_type p, size_t i) const noexcept {
        return p + i;
    }
};

// Version 1: raw
void vadd(int n, const double* a, const double* b, double* c) {
    for (int i = 0; i < n; i++) {
        c[i] = a[i] + b[i];
    }
}

// Version 2: raw + restrict
void vadd(int n, const double *__restrict__ a, const double *__restrict__ b, double *__restrict__ c) {
    for (int i = 0; i < n; i++) {
        c[i] = a[i] + b[i];
    }
}

// Version 3: mdspan
void vadd(int n, const double *_a, const double *_b, double *_c) {
    stdex::mdspan<const double, stdex::dextents<size_t, 1>> a(_a, n);
    stdex::mdspan<const double, stdex::dextents<size_t, 1>> b(_b, n);
    stdex::mdspan<double, stdex::dextents<size_t, 1>> c(_c, n);
    for (int i = 0; i < n; i++) {
        c[i] = a[i] + b[i];
    }
}

// Version 4: mdspan + restrict
void vadd(int n, const double *_a, const double *_b, double *_c) {
    stdex::mdspan<const double, stdex::dextents<size_t, 1>, stdex::layout_right, restrict_accessor<const double>> a(_a, n);
    stdex::mdspan<const double, stdex::dextents<size_t, 1>, stdex::layout_right, restrict_accessor<const double>> b(_b, n);
    stdex::mdspan<double, stdex::dextents<size_t, 1>, stdex::layout_right, restrict_accessor<double>> c(_c, n);
    for (int i = 0; i < n; i++) {
        c[i] = a[i] + b[i];
    }
}

int main() {
    const int n = 8;
    double a[n] = {0, 1, 2, 3, 4, 5, 6, 7};
    double b[n] = {1, 2, 3, 4, 5, 6, 7, 8};
    double c[n];
    vadd(n, a, b, c);
    for (int i = 0; i < n; i++) {
        std::cout << c[i] << " ";
    }
    std::cout << std::endl;
    return 0;
}
```

I tried all 4 versions, compiled with `-O3 -march=native -std=c++2b`. Version 1, 3 and 4 produced identical code. Only Version 2 produced different code, with less instructions (I have look at the asm, it seems like for Version 1, 3, and 4 it determine whether the pointers are aliases at runtime). It seems g++ failed to recognize `restrict` in `restrict_accessor`.

Maybe g++ is not capable enough, but still thanks. Feel free to close the issue.

**Comment by mhoemmen** [comment]
@roastduck Thanks for trying this out and reporting results!  Regarding Version 4 (mdspan + restrict), I don't think that is the most idiomatic use of `restrict`.  The idiomatic way is to mark function parameters as restrict, not to apply restrict to non-restrict pointers in the same block.  For example:
```c++
template<class ElementType>
using restrict_array_view =
    stdex::mdspan<ElementType, stdex::dextents<size_t, 1>, stdex::layout_right, restrict_accessor<ElementType>>;

void vadd(int n, restrict_array_view<const double> a, restrict_array_view<const double> b, restrict_array_view<double> c) {
    for (int i = 0; i < n; i++) {
        c[i] = a[i] + b[i];
    }
}

void function_calling_vadd(int n, const double* a, const double* b, double* c) {
    vadd(n, restrict_array_view{a, n}, restrict_array_view{b, n}, restrict_array_view{c, n});
}
```
Please see Examples 2 and 5 in Section 6.7.3.1 ("Formal definition of `restrict`") of the current [C Standard working draft](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2596.pdf).

That being said, I haven't done this experiment myself.  Also, as you pointed out, GCC might have troubles doing the alias analysis through the extra layers of `mdspan`.  The Kokkos team has certainly seen that issue with other compilers.

Thanks again for reporting results!  I'm grateful for these extra data.  Please feel free to reopen the issue or open a new one if you have any questions.

**Comment by roastduck** [comment]
> @roastduck Thanks for trying this out and reporting results! Regarding Version 4 (mdspan + restrict), I don't think that is the most idiomatic use of `restrict`. The idiomatic way is to mark function parameters as restrict, not to apply restrict to non-restrict pointers in the same block. For example:
> 
> ```c++
> template<class ElementType>
> using restrict_array_view =
>     stdex::mdspan<ElementType, stdex::dextents<size_t, 1>, stdex::layout_right, restrict_accessor<ElementType>>;
> 
> void vadd(int n, restrict_array_view<const double> a, restrict_array_view<const double> b, restrict_array_view<double> c) {
>     for (int i = 0; i < n; i++) {
>         c[i] = a[i] + b[i];
>     }
> }
> 
> void function_calling_vadd(int n, const double* a, const double* b, double* c) {
>     vadd(n, restrict_array_view{a, n}, restrict_array_view{b, n}, restrict_array_view{c, n});
> }
> ```

Just FYI, I tried this implementation and it successfully produced optimized code.
